File: src/client/java/net/greenjab/fixedminecraft/ClientSyncHandler.java
package net.greenjab.fixedminecraft;

import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.greenjab.fixedminecraft.screens.MapBookScreen;
import net.greenjab.fixedminecraft.network.MapBookOpenPayload;
import net.greenjab.fixedminecraft.network.MapBookSyncPayload;
import net.greenjab.fixedminecraft.network.MapPositionPayload;
import net.greenjab.fixedminecraft.network.SaturationSyncPayload;
import net.greenjab.fixedminecraft.network.TrainPayload;
import net.greenjab.fixedminecraft.network.VillagerNeedsPayload;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookState;
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookStateManager;
import net.greenjab.fixedminecraft.registry.item.map_book.MapStateAccessor;
import net.greenjab.fixedminecraft.registry.other.FixedFurnaceMinecartEntity;
import net.greenjab.fixedminecraft.registry.registries.ParticleRegistry;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientLevel;
import net.minecraft.core.particles.SimpleParticleType;
import net.minecraft.network.chat.Component;
import net.minecraft.util.RandomSource;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.npc.villager.Villager;
import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
import net.minecraft.world.phys.Vec3;

import java.util.ArrayList;

/** Credit: Nettakrim, Squeek502, Bawnorton */
public class ClientSyncHandler {
    public static void init() {
        ClientPlayNetworking.registerGlobalReceiver(SaturationSyncPayload.ID, (payload, context) ->
                context.client().execute(() ->
                        context.client().player.getFoodData().setSaturation(payload.getSaturation())));

        ClientPlayNetworking.registerGlobalReceiver(MapBookOpenPayload.PACKET_ID, ClientSyncHandler::mapBookOpen);
        ClientPlayNetworking.registerGlobalReceiver(MapBookSyncPayload.PACKET_ID, ClientSyncHandler::mapBookSync);
        ClientPlayNetworking.registerGlobalReceiver(MapPositionPayload.PACKET_ID, ClientSyncHandler::mapPosition);
        ClientPlayNetworking.registerGlobalReceiver(TrainPayload.PACKET_ID, ClientSyncHandler::train);
        ClientPlayNetworking.registerGlobalReceiver(VillagerNeedsPayload.PACKET_ID, ClientSyncHandler::villagerNeed);

    }
    private static void mapBookOpen(MapBookOpenPayload payload, ClientPlayNetworking.Context context) {
        context.client().execute(() -> context.client().setScreen(new MapBookScreen(payload.itemStack())));
    }

    private static void mapBookSync(MapBookSyncPayload payload, ClientPlayNetworking.Context context) {
        if (payload.mapIDs().length > 0) {
            context.client().execute(() -> {
                ArrayList<Integer> ints = new ArrayList<>();
                for (int i = 0; i < payload.mapIDs().length;i++) {
                    ints.add(payload.mapIDs()[i]);
                }
                MapBookStateManager.INSTANCE.putClientMapBookState(
                        payload.bookID(),

                        new MapBookState(ints, payload.players(), payload.marker())
                );
            });
        }
    }

    private static void mapPosition(MapPositionPayload payload, ClientPlayNetworking.Context context) {
        context.client().execute(()-> {
            ClientLevel level = context.client().level;
            if (level != null) {
                MapItemSavedData mapstate = level.getMapData(payload.mapIdComponent());
                if (mapstate != null) {
                    ((MapStateAccessor)mapstate).fixedminecraft$setPosition(payload.centerX(), payload.centerZ());
                }
            }
        });
    }

    private static void train(TrainPayload payload, ClientPlayNetworking.Context context) {
        context.client().execute(()-> {
            ClientLevel level = context.client().level;
            if (level != null) {
                Entity entity = level.getEntity((payload.train().getFirst()));
                if (entity instanceof FixedFurnaceMinecartEntity furnaceMinecart) {
                    furnaceMinecart.setTrain(payload.train());
                }
            }
        });
    }

    private static void villagerNeed(VillagerNeedsPayload payload, ClientPlayNetworking.Context context) {
        context.client().execute(()-> {
            ClientLevel level = context.client().level;
            if (level != null) {
                Entity entity = level.getEntity((payload.villager()));
                if (entity instanceof Villager villager) {
                    RandomSource random = villager.getRandom();
                    Minecraft.getInstance().player.sendSystemMessage(Component.translatable("entity.fixedminecraft.villager."+payload.need(), villager.getName()));

                    SimpleParticleType particle = switch (payload.need()) {
                        case "hungry","very_hungry" -> ParticleRegistry.VILLAGER_HUNGRY;
                        case "tired","very_tired" -> ParticleRegistry.VILLAGER_TIRED;
                        case "lonely","very_lonely" -> ParticleRegistry.VILLAGER_LONELY;
                        case "lazy","very_lazy" -> ParticleRegistry.VILLAGER_LAZY;
                        case "dark","very_dark" -> ParticleRegistry.VILLAGER_DARK;
                        case "night" -> ParticleRegistry.VILLAGER_NIGHT;
                        default -> null;
                    };
                    if (particle != null) {
                        Vec3 d = Minecraft.getInstance().player.position().subtract(villager.position()).horizontal().normalize();
                        level.addAlwaysVisibleParticle(
                                particle,
                                true,
                                villager.getX() + d.x*0.5 + random.nextDouble() / 5.0 * (random.nextBoolean() ? 1 : -1),
                                villager.getY() + 1 + random.nextDouble() / 2.0,
                                villager.getZ() + d.z*0.5 + random.nextDouble() / 5.0 * (random.nextBoolean() ? 1 : -1),
                                0.0,
                                0.02,
                                0.0
                        );
                    }
                }
            }
        });
    }
}

File: src/client/kotlin/net/greenjab/fixedminecraft/network/ClientSyncHandler.kt
package net.greenjab.fixedminecraft.network

import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.greenjab.fixedminecraft.map_book.MapBookScreen
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookState
import net.greenjab.fixedminecraft.registry.item.map_book.MapBookStateManager
import net.greenjab.fixedminecraft.registry.item.map_book.MapStateAccessor
import net.greenjab.fixedminecraft.util.ExhaustionHelper
import net.minecraft.block.Block
import net.minecraft.entity.player.PlayerEntity


object ClientSyncHandler {
    @Environment(EnvType.CLIENT)
    fun init() {
        ClientPlayNetworking.registerGlobalReceiver(
            ExhaustionSyncPayload.ID
        ) { payload: ExhaustionSyncPayload, context: ClientPlayNetworking.Context ->
            context.client().execute {
                ExhaustionHelper.setExhaustion(context.client().player as PlayerEntity, payload.exhaustion)
            }
        }
        ClientPlayNetworking.registerGlobalReceiver(
            SaturationSyncPayload.ID
        ) { payload: SaturationSyncPayload, context: ClientPlayNetworking.Context ->
            context.client().execute {
                context.client().player!!.hungerManager.saturationLevel = payload.saturation
            }
        }

        ClientPlayNetworking.registerGlobalReceiver(MapBookOpenPayload.PACKET_ID, ClientSyncHandler::mapBookOpen);
        ClientPlayNetworking.registerGlobalReceiver(MapBookSyncPayload.PACKET_ID, ClientSyncHandler::mapBookSync);
         ClientPlayNetworking.registerGlobalReceiver(MapPositionPayload.PACKET_ID, ClientSyncHandler::mapPosition);

         ClientPlayNetworking.registerGlobalReceiver(BookShelfSyncPayload.PACKET_ID, ClientSyncHandler::bookShelfSync);

    }
    private fun mapBookOpen(payload: MapBookOpenPayload, context: ClientPlayNetworking.Context) {
        context.client().execute { context.client().setScreen(MapBookScreen(payload.itemStack)) }
    }

    private fun mapBookSync(payload: MapBookSyncPayload, context: ClientPlayNetworking.Context) {
        if (payload.mapIDs.size > 0) {
            context.client().execute {
                MapBookStateManager.INSTANCE.putClientMapBookState(
                    payload.bookID,
                    MapBookState(payload.mapIDs, payload.players)
                )
            }
        }
    }

    private fun mapPosition(payload: MapPositionPayload, context: ClientPlayNetworking.Context) {
        context.client().execute {
            val world = context.client().world
            if (world != null) {
                val mapstate = world.getMapState(payload.mapIdComponent)
                if (mapstate != null) {
                    (mapstate as MapStateAccessor).`fixed$setPosition`(payload.centerX, payload.centerZ)
                }
            }
        }
    }

    private fun bookShelfSync(payload: BookShelfSyncPayload, context: ClientPlayNetworking.Context) {

        val pos = payload.pos
        context.client().execute {
            assert(context.client().world != null)
            context.client().world!!.updateListeners(
                pos,
                context.client().world!!.getBlockState(pos),
                context.client().world!!.getBlockState(pos),
                Block.NOTIFY_LISTENERS
            )
        }
    }
}
